# Crash on Wayland: GLFW error callback treats non-fatal `GLFW_FEATURE_UNAVAILABLE` (window position) as fatal

# Crash on Wayland: GLFW error callback treats non-fatal `GLFW_FEATURE_UNAVAILABLE` (window position) as fatal

## Summary
On a native **Wayland** session, the game cannot start. `blaze3d` installs a GLFW
error callback *before* init (`com.mojang.blaze3d.platform.GLX._initGlfw`) that
rethrows **any** GLFW error as a fatal `IllegalStateException`. The Wayland
backend legitimately reports `GLFW_FEATURE_UNAVAILABLE (0x1000C)` for window
position — which is **non-fatal by the GLFW spec** (Wayland has no concept of
absolute window position) — and this aborts startup.

## Environment
- OS: Fedora Linux 44 (kernel 7.0.10-201.fc44.x86_64)
- Session: **Wayland** (`XDG_SESSION_TYPE=wayland`), XWayland available on `:0`
- GPU: hybrid Intel UHD 630 (i7-9850H) + NVIDIA Quadro P620
- Java: Microsoft OpenJDK 25.0.1
- Minecraft: 26.1.2
- Forge: 26.1.2-64.0.8 (also reproduced on 64.0.4)
- LWJGL: 3.4.1+2  (single `libglfw.so` compiled with **both** X11 and Wayland backends)
- `GFLW Platform: wayland`

## Stack trace (main crash, native Wayland)
```
java.lang.IllegalStateException: GLFW error before init: [0x1000C]Wayland: The platform does not provide the window position
	at com.mojang.blaze3d.platform.GLX.lambda$_initGlfw$0(GLX.java:43)
	at com.mojang.blaze3d.platform.Window.checkGlfwError(Window.java:179)
	at com.mojang.blaze3d.platform.GLX._initGlfw(GLX.java:42)
	at com.mojang.blaze3d.systems.RenderSystem.initBackendSystem(RenderSystem.java:179)
	at net.minecraft.client.Minecraft.<init>(Minecraft.java:463)
	at net.minecraft.client.main.Main.main(Main.java:223)
```

## Steps to reproduce
1. Boot into a Wayland session (GNOME/KDE Wayland).
2. Launch Minecraft 26.1.2 with Forge 26.1.2-64.0.x.
3. Crash during "Initializing game" with the trace above.

## Secondary issue (when forcing X11 fallback)
Trying to dodge the bug by making the Wayland backend unavailable
(`WAYLAND_DISPLAY=` invalid) does **not** fall back to X11 cleanly — Forge's
early display aborts:
```
Failed to create wl_display (No such file or directory)
Caused by: java.lang.IllegalStateException: Unable to initialize GLFW
	at net.minecraftforge.fml.earlydisplay.DisplayWindow.initWindow(DisplayWindow.java:344)
```
(GLFW's X11 backend additionally needs `XAUTHORITY`/`~/.Xauthority` for the
XWayland cookie, which the launch environment does not always provide.)

## Expected behavior
A non-fatal `GLFW_FEATURE_UNAVAILABLE` for window position on Wayland should be
ignored (logged, not thrown). The game should start on native Wayland, as it
does on X11/XWayland.

## Workaround currently used
Force the game onto X11/XWayland: launch with the Wayland backend disabled and a
valid X11 cookie present (`~/.Xauthority`), so GLFW selects the X11 backend where
`glfwGetWindowPos` is supported.

## Notes
- Root cause is in `blaze3d` (`GLX._initGlfw` error callback rethrowing
  `GLFW_FEATURE_UNAVAILABLE`); may also warrant a fix upstream in Mojang.
- Reproducible on a clean profile (no coremods needed); mods are not involved —
  the crash precedes mod window creation.

--- a/fmlearlydisplay/src/main/java/net/minecraftforge/fml/earlydisplay/DisplayWindow.java
+++ b/fmlearlydisplay/src/main/java/net/minecraftforge/fml/earlydisplay/DisplayWindow.java
@@ -459,6 +459,7 @@ public class DisplayWindow implements ImmediateWindowProvider {
         this.winWidth = x[0];
         this.winHeight = y[0];
 
+        // Setting the window position isn't supported on wayland, so check the error here
         glfwSetWindowPos(window, (vidmode.width() - this.winWidth) / 2 + monitorX, (vidmode.height() - this.winHeight) / 2 + monitorY);
         handleLastGLFWError();
 
@@ -483,8 +484,10 @@ public class DisplayWindow implements ImmediateWindowProvider {
 
         // Show the window
         glfwShowWindow(window);
+        // Getting the window position isn't supported on wayland, so check the error here
         glfwGetWindowPos(window, x, y);
         handleLastGLFWError();
+        
         this.winX = x[0];
         this.winY = y[0];
         glfwGetFramebufferSize(window, x, y);

--- a/fmlearlydisplay/src/main/java/net/minecraftforge/fml/earlydisplay/DisplayWindow.java
+++ b/fmlearlydisplay/src/main/java/net/minecraftforge/fml/earlydisplay/DisplayWindow.java
@@ -460,6 +460,7 @@ public class DisplayWindow implements ImmediateWindowProvider {
         this.winHeight = y[0];
 
         glfwSetWindowPos(window, (vidmode.width() - this.winWidth) / 2 + monitorX, (vidmode.height() - this.winHeight) / 2 + monitorY);
+        handleLastGLFWError();
 
         // Attempt setting the icon
 //        int[] channels = new int[1];
@@ -483,12 +484,25 @@ public class DisplayWindow implements ImmediateWindowProvider {
         // Show the window
         glfwShowWindow(window);
         glfwGetWindowPos(window, x, y);
+        handleLastGLFWError();
         this.winX = x[0];
         this.winY = y[0];
         glfwGetFramebufferSize(window, x, y);
         this.fbWidth = x[0];
         this.fbHeight = y[0];
         glfwPollEvents();
+    }
+
+    private static void handleLastGLFWError() {
+        handleLastGLFWError((error, description) -> {
+            if (error == GLFW_FEATURE_UNAVAILABLE) {
+                // suppress window pos errors for unsupported platforms (wayland)
+                LOGGER.debug(String.format("Suppressing GLFW error: [0x%X]%s", error, description));
+                return;
+            }
+
+            throw new IllegalStateException(String.format("GLFW error: [0x%X]%s", error, description));
+        });
     }
 
     private void winResize(long window, int width, int height) {

